# Overview - Identify and Remove Duplicates

`4/20/2025` update of the notebook. I have done enough verifications in the past that I am going to do an abbreviated version of the code here

This was originally an update of this notebook:
- *IRS Form 990 e-File Data (8a) -- Identify and Drop Duplicate Org-Year Filings*

Notes 4/20/2025
```python
- df['direct_donations'] = df['F9_08_PC_TOTAL_CONTRIBUTIONS']-df['F9_08_PC_FEDERATED_CAMPAIGNS'] -df['F9_08_PC_GOVERNMENT_GRANTS']
```

In [1]:
import numpy as np
import pandas as pd
from pandas import DataFrame
from pandas import Series

In [2]:
print(pd.__version__)

2.2.2


In [3]:
from platform import python_version
print(python_version())

3.10.11


In [4]:
# http://pandas.pydata.org/pandas-docs/stable/options.html
pd.set_option('display.max_columns', None)
# http://pandas.pydata.org/pandas-docs/stable/options.html
pd.set_option('display.max_colwidth', 4000)

In [5]:
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [6]:
pd.options.display.float_format = '{:,.1f}'.format

In [7]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)
warnings.simplefilter(action='ignore', category=pd.errors.SettingWithCopyWarning)

In [8]:
import datetime
import gc

#### Set working directory

In [9]:
cd "C:\\Users\\Gregory\\IRS 990 Control Variables\\"

C:\Users\Gregory\IRS 990 Control Variables


# Read in PANDAS DF

In [9]:
#%%time
#import datetime
#print ("Current date and time : ", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"), '\n')
#df = pd.read_pickle('990 and BMF control variables for all NEW filings February 2024 -- 277,112 501c3 orgs (N=664,761).pkl.gz', compression='gzip')
#print('# of columns:', len(df.columns))
#print('# of observations:', len(df))
#df[:1]

Current date and time :  2024-06-16 14:51:11 

# of columns: 340
# of observations: 664761
CPU times: total: 2min 33s
Wall time: 2min 46s


URL  \
812471  https://s3.amazonaws.com/irs-form-990/202323189349305317_public.xml   

        F9_09_PC_FEES_FOR_SVCE_FR_TOT F9_00_HD_BUILD_TIME_STAMP fiscal_year  \
812471                            0.0 2023-04-26 12:10:37+00:00        2022   

              EIN                                            BusinessName  \
812471  010017496  {'BusinessNameLine1Txt': 'AGAMENTICUS YACHT CLUB INC'}   

       BusinessNameControlTxt    PhoneNum  \
812471                   AGAM  2073638510   

                                                                                                        USAddress  \
812471  {'AddressLine1Txt': 'PO BOX 534', 'CityNm': 'YORK HARBOR', 'StateAbbreviationCd': 'ME', 'ZIPCd': '03911'}   

       InCareOfNm ForeignAddress ForeignPhoneNum  F9_00_HD_ADDR_CHANGE  \
812471        NaN            NaN             NaN                   0.0   

        F9_00_HD_AMENDED_RETURN F9_00_HD_CTRY_OF_DOMICILE  \
812471                      0.0                       NaN   

        F9_00_HD_EXEMPT_STATUS_4847A1  F9_00_HD_EXEMPT_STATUS_501C  \
812471                            0.0                          NaN   

        F9_00_HD_EXEMPT_STATUS_501C3  F9_00_HD_FINAL_RETURN  \
812471                           1.0                    0.0   

       F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  F9_00_HD_GROUP_RETURN  \
812471                       NaN               376800                      0   

        F9_00_HD_INCLUDES_SUBORD_ORGS  F9_00_HD_INITIAL_RETURN  \
812471                            0.0                      0.0   

       F9_00_HD_PRIN_OFF_NAME F9_00_HD_SIGNING_OFFICER_SIGNTR  \
812471            DANIEL FORD                      2023-11-13   

       F9_00_HD_SPECIAL_CONDITION_DESC F9_00_HD_STATE_OF_DOMICILE  \
812471                             NaN                         ME   

       F9_00_HD_TAX_PER_BEGIN F9_00_HD_TAX_PER_END  F9_00_HD_TAX_YEAR  \
812471             2022-01-01           2022-12-31               2022   

              F9_00_HD_TIME_STAMP  F9_00_HD_TYPE_ORG_ASSOCIATION  \
812471  2023-11-14 10:30:26-06:00                            0.0   

        F9_00_HD_TYPE_ORG_CORP  F9_00_HD_TYPE_ORG_OTHER  \
812471                     1.0                      0.0   

       F9_00_HD_TYPE_ORG_OTHER_DESC  F9_00_HD_TYPE_ORG_TRUST F9_00_HD_WEBSITE  \
812471                          NaN                      0.0  WWW.AYCSAIL.ORG   

        F9_00_HD_YEAR_FORMED  F9_01_PC_BEN_PAID_MEMB_PRIOR  \
812471               1,937.0                           0.0   

        F9_01_PC_CONTR_GRANTS_CURR  F9_01_PC_CONTR_GRANTS_PRIOR  \
812471                      279970                          0.0   

        F9_01_PC_GRANTS_PRIOR  F9_01_PC_INDEP_VOTING_MEMB  \
812471                    0.0                          13   

        F9_01_PC_INVEST_INCOME_PRIOR  F9_01_PC_NET_ASSETS_BOY  \
812471                           0.0                273,331.0   

        F9_01_PC_OTHER_EXPENSE_PRIOR  F9_01_PC_OTHER_REV_PRIOR  \
812471                           0.0                       0.0   

        F9_01_PC_PROF_FUNDRISING_EXP_CURR  F9_01_PC_PROF_FUNDRISING_EXP_PRIOR  \
812471                                  0                                 0.0   

        F9_01_PC_PROG_SERVICE_REV_PRIOR  F9_01_PC_REV_LESS_EXP_CURR  \
812471                              0.0                      184620   

        F9_01_PC_REV_LESS_EXP_PRIOR  F9_01_PC_TERMINATION_CONTRACTION  \
812471                          0.0                               0.0   

        F9_01_PC_TOT_ASSETS_EOY  F9_01_PC_TOT_EXP_PRIOR  \
812471                   413907                     0.0   

        F9_01_PC_TOT_FNDR_EXP_CURR  F9_01_PC_TOT_INDIV_EMPLOYED  \
812471                        2744                           16   

        F9_01_PC_TOT_INDIV_VOLUNTEERS  F9_01_PC_TOT_LIABILITIES_EOY  \
812471                           20.0                          8282   

        F9_01_PC_TOT_REVENUE_PRIOR  F9_01_PC_TOT_UBI_GROSS  \
812471                         0.0          

In [10]:
%%time
print ("Current date and time : ", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"), '\n')
df = pd.read_feather('D:/990_and_bmf_april_2025_all_controls_351875_orgs_2633270_filings.feather')
print('# of columns:', len(df.columns))
print('# of observations:', len(df))
df[:1]

Current date and time :  2025-04-20 19:40:56 

# of columns: 356
# of observations: 2633270
CPU times: total: 57.2 s
Wall time: 33.9 s


_id OrganizationName  \
3016720  65c1a1d52a9ba8ce45342904             None   

                                                                         URL  \
3016720  https://s3.amazonaws.com/irs-form-990/202323189349305317_public.xml   

          DLN TaxPeriod  F9_09_PC_FEES_FOR_SVCE_FR_TOT  \
3016720  None      None                              0   

        F9_00_HD_BUILD_TIME_STAMP fiscal_year        EIN  Name NameControl  \
3016720 2023-04-26 12:10:37+00:00        2022  010017496  None        None   

        Phone  \
3016720  None   

                                                                                                                                                                                                                            USAddress  \
3016720  {'AddressLine1': None, 'AddressLine1Txt': 'PO BOX 534', 'AddressLine2': None, 'AddressLine2Txt': None, 'City': None, 'CityNm': 'YORK HARBOR', 'State': None, 'StateAbbreviationCd': 'ME', 'ZIPCd': '03911', 'ZIPCode': None}   

        ForeignAddress InCareOfName  \
3016720           None         None   

                                                   BusinessName  \
3016720  {'BusinessNameLine1Txt': 'AGAMENTICUS YACHT CLUB INC'}   

        BusinessNameControlTxt    PhoneNum InCareOfNm ForeignPhoneNum  \
3016720                   AGAM  2073638510       None            None   

         F9_00_HD_ADDR_CHANGE  F9_00_HD_AMENDED_RETURN  \
3016720                     0                        0   

        F9_00_HD_CTRY_OF_DOMICILE  F9_00_HD_EXEMPT_STATUS_4847A1  \
3016720                      None                              0   

         F9_00_HD_EXEMPT_STATUS_501C  F9_00_HD_EXEMPT_STATUS_501C3  \
3016720                         <NA>                             1   

         F9_00_HD_FINAL_RETURN F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  \
3016720                      0                      None               376800   

         F9_00_HD_GROUP_RETURN  F9_00_HD_INCLUDES_SUBORD_ORGS  \
3016720                      0                              0   

         F9_00_HD_INITIAL_RETURN F9_00_HD_PRIN_OFF_NAME  \
3016720                        0            DANIEL FORD   

        F9_00_HD_SIGNING_OFFICER_SIGNTR F9_00_HD_SPECIAL_CONDITION_DESC  \
3016720                      2023-11-13                            None   

        F9_00_HD_STATE_OF_DOMICILE F9_00_HD_TAX_PER_BEGIN  \
3016720                         ME             2022-01-01   

        F9_00_HD_TAX_PER_END  F9_00_HD_TAX_YEAR       F9_00_HD_TIME_STAMP  \
3016720           2022-12-31               2022 2023-11-14 16:30:26+00:00   

         F9_00_HD_TYPE_ORG_ASSOCIATION  F9_00_HD_TYPE_ORG_CORP  \
3016720                              0                       1   

         F9_00_HD_TYPE_ORG_OTHER F9_00_HD_TYPE_ORG_OTHER_DESC  \
3016720                        0                         None   

         F9_00_HD_TYPE_ORG_TRUST F9_00_HD_WEBSITE  F9_00_HD_YEAR_FORMED  \
3016720                        0  WWW.AYCSAIL.ORG                  1937   

         F9_01_PC_BEN_PAID_MEMB_PRIOR  F9_01_PC_CONTR_GRANTS_CURR  \
3016720                             0                      279970   

         F9_01_PC_CONTR_GRANTS_PRIOR  F9_01_PC_GRANTS_PRIOR  \
3016720                            0                      0   

         F9_01_PC_INDEP_VOTING_MEMB  F9_01_PC_INVEST_INCOME_PRIOR  \
3016720                          13                             0   

         F9_01_PC_NET_ASSETS_BOY  F9_01_PC_OTHER_EXPENSE_PRIOR  \
3016720                   273331                             0   

         F9_01_PC_OTHER_REV_PRIOR  F9_01_PC_PROF_FUNDRISING_EXP_CURR  \
3016720                         0                                  0   

         F9_01_PC_PROF_FUNDRISING_EXP_PRIOR  F9_01_PC_PROG_SERVICE_REV_PRIOR  \
3016720                                   0                                0   

         F9_01_PC_REV_LESS_EXP_CURR  F9_01_PC_REV_LESS_EXP_PRIOR  \
3016720                      184620                        

In [13]:
print('# of observations:', len(df))
print('# of unique organizations:', len(set(df['EIN'].tolist())))
print('# of observations that are by 501(c)(3) organizations:', df['501c3'].value_counts())

# of observations: 2633270
# of unique organizations: 351875
# of observations that are by 501(c)(3) organizations: 501c3
1    2633270
Name: count, dtype: int64


In [15]:
pd.set_option('display.float_format', lambda x: '%.2f' % x)

# Select Latest Filing per Org/Year

In [16]:
df['F9_00_HD_TAX_YEAR'].value_counts().sort_index()

F9_00_HD_TAX_YEAR
2009     28392
2010     92910
2011    120596
2012    135811
2013    150462
2014    165382
2015    176852
2016    184814
2017    193481
2018    202483
2019    216427
2020    242299
2021    256105
2022    264605
2023    200213
2024      2438
Name: count, dtype: Int64

In [14]:
df.sample(1)

URL  \
528959  https://s3.amazonaws.com/irs-form-990/202310169349300311_public.xml   

        F9_09_PC_FEES_FOR_SVCE_FR_TOT F9_00_HD_BUILD_TIME_STAMP fiscal_year  \
528959                           0.00 2023-03-07 01:45:13+00:00        2022   

              EIN                                            BusinessName  \
528959  376039121  {'BusinessNameLine1Txt': 'Prairieland United Way Inc'}   

       BusinessNameControlTxt    PhoneNum  \
528959                   PRAI  2172454557   

                                                                                                                        USAddress  \
528959  {'AddressLine1Txt': '200 W Douglas Ave Box 244', 'CityNm': 'Jacksonville', 'StateAbbreviationCd': 'IL', 'ZIPCd': '62651'}   

       InCareOfNm ForeignAddress ForeignPhoneNum  F9_00_HD_ADDR_CHANGE  \
528959        NaN            NaN             NaN                  0.00   

        F9_00_HD_AMENDED_RETURN F9_00_HD_CTRY_OF_DOMICILE  \
528959                     0.00                       NaN   

        F9_00_HD_EXEMPT_STATUS_4847A1  F9_00_HD_EXEMPT_STATUS_501C  \
528959                           0.00                          NaN   

        F9_00_HD_EXEMPT_STATUS_501C3  F9_00_HD_FINAL_RETURN  \
528959                          1.00                   0.00   

       F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  F9_00_HD_GROUP_RETURN  \
528959                       NaN               485118                      0   

        F9_00_HD_INCLUDES_SUBORD_ORGS  F9_00_HD_INITIAL_RETURN  \
528959                           0.00                     0.00   

       F9_00_HD_PRIN_OFF_NAME F9_00_HD_SIGNING_OFFICER_SIGNTR  \
528959           Karen Walker                      2023-01-09   

       F9_00_HD_SPECIAL_CONDITION_DESC F9_00_HD_STATE_OF_DOMICILE  \
528959                             NaN                         IL   

       F9_00_HD_TAX_PER_BEGIN F9_00_HD_TAX_PER_END  F9_00_HD_TAX_YEAR  \
528959             2021-07-01           2022-06-30               2021   

              F9_00_HD_TIME_STAMP  F9_00_HD_TYPE_ORG_ASSOCIATION  \
528959  2023-01-16 08:47:42-06:00                           0.00   

        F9_00_HD_TYPE_ORG_CORP  F9_00_HD_TYPE_ORG_OTHER  \
528959                    1.00                     0.00   

       F9_00_HD_TYPE_ORG_OTHER_DESC  F9_00_HD_TYPE_ORG_TRUST  \
528959                          NaN                     0.00   

                    F9_00_HD_WEBSITE  F9_00_HD_YEAR_FORMED  \
528959  www.prairielandunitedway.org               1959.00   

        F9_01_PC_BEN_PAID_MEMB_PRIOR  F9_01_PC_CONTR_GRANTS_CURR  \
528959                          0.00                      478504   

        F9_01_PC_CONTR_GRANTS_PRIOR  F9_01_PC_GRANTS_PRIOR  \
528959                    453837.00              351004.00   

        F9_01_PC_INDEP_VOTING_MEMB  F9_01_PC_INVEST_INCOME_PRIOR  \
528959                          19                       1693.00   

        F9_01_PC_NET_ASSETS_BOY  F9_01_PC_OTHER_EXPENSE_PRIOR  \
528959                294621.00                      31938.00   

        F9_01_PC_OTHER_REV_PRIOR  F9_01_PC_PROF_FUNDRISING_EXP_CURR  \
528959                      0.00                                  0   

        F9_01_PC_PROF_FUNDRISING_EXP_PRIOR  F9_01_PC_PROG_SERVICE_REV_PRIOR  \
528959                                0.00                             0.00   

        F9_01_PC_REV_LESS_EXP_CURR  F9_01_PC_REV_LESS_EXP_PRIOR  \
528959                       26521                     -4017.00   

        F9_01_PC_TERMINATION_CONTRACTION  F9_01_PC_TOT_ASSETS_EOY  \
528959                              0.00                   323154   

        F9_01_PC_TOT_EXP_PRIOR  F9_01_PC_TOT_FNDR_EXP_CURR  \
528959               459547.00                       15073   

        F9_01_PC_TOT_INDIV_EMPLOYED  F9_01_PC_TOT_INDIV_VOLUNTEERS  \
528959                            2                          70.00   

        F9_01_PC_TOT_LIABILITIES_EOY  F9_01_PC_TOT_REVENUE_PRIOR  \
528959                          7153    

#### Create DF with duplicate observations

In [17]:
%%time
grouped = df.groupby(['EIN', 'F9_00_HD_TAX_YEAR']).count()
print(len(grouped[grouped['URL']>1]))
print(grouped[grouped['URL']>1]['URL'].value_counts())
grouped = grouped[grouped['URL']>1]
grouped = grouped.reset_index()
grouped = grouped[['EIN', 'F9_00_HD_TAX_YEAR', 'URL']]
grouped.columns = ['EIN', 'F9_00_HD_TAX_YEAR', 'number of org-year duplicates']
print(len(grouped))
grouped[:2]

33668
URL
2    32664
3      899
4       90
5       14
6        1
Name: count, dtype: int64
33668
CPU times: total: 13.4 s
Wall time: 14 s


EIN  F9_00_HD_TAX_YEAR  number of org-year duplicates
0  010211481               2017                              2
1  010211486               2015                              2

In [25]:
%%time
pd.merge(df, grouped, left_on=['EIN', 'F9_00_HD_TAX_YEAR'], right_on=['EIN', 'F9_00_HD_TAX_YEAR'], how='left', indicator=True)['_merge'].value_counts()

CPU times: total: 22 s
Wall time: 23.1 s


_merge
left_only     2564809
both            68461
right_only          0
Name: count, dtype: int64

In [15]:
#print((14159*2)+(366*3)+(19*4)+(2*5))
#print(grouped['number of org-year duplicates'].sum())

29502
29502


In [18]:
%%time
print('# of observations:', len(df))
print('# of columns:', len(df.columns))
print(len(pd.merge(df, grouped, left_on=['EIN', 'F9_00_HD_TAX_YEAR'], right_on=['EIN', 'F9_00_HD_TAX_YEAR'], how='left', indicator=True)))
df = pd.merge(df, grouped, left_on=['EIN', 'F9_00_HD_TAX_YEAR'], right_on=['EIN', 'F9_00_HD_TAX_YEAR'], how='left', indicator=True)
print(df['_merge'].value_counts())
df = df.drop('_merge', axis=1)
print('# of observations:', len(df))
print('# of columns:', len(df.columns))
df[:1]

# of observations: 2633270
# of columns: 356
2633270
_merge
left_only     2564809
both            68461
right_only          0
Name: count, dtype: int64
# of observations: 2633270
# of columns: 357
CPU times: total: 45 s
Wall time: 46.5 s


_id OrganizationName  \
0  65c1a1d52a9ba8ce45342904             None   

                                                                   URL   DLN  \
0  https://s3.amazonaws.com/irs-form-990/202323189349305317_public.xml  None   

  TaxPeriod  F9_09_PC_FEES_FOR_SVCE_FR_TOT F9_00_HD_BUILD_TIME_STAMP  \
0      None                              0 2023-04-26 12:10:37+00:00   

  fiscal_year        EIN  Name NameControl Phone  \
0        2022  010017496  None        None  None   

                                                                                                                                                                                                                      USAddress  \
0  {'AddressLine1': None, 'AddressLine1Txt': 'PO BOX 534', 'AddressLine2': None, 'AddressLine2Txt': None, 'City': None, 'CityNm': 'YORK HARBOR', 'State': None, 'StateAbbreviationCd': 'ME', 'ZIPCd': '03911', 'ZIPCode': None}   

  ForeignAddress InCareOfName  \
0           None         None   

                                             BusinessName  \
0  {'BusinessNameLine1Txt': 'AGAMENTICUS YACHT CLUB INC'}   

  BusinessNameControlTxt    PhoneNum InCareOfNm ForeignPhoneNum  \
0                   AGAM  2073638510       None            None   

   F9_00_HD_ADDR_CHANGE  F9_00_HD_AMENDED_RETURN F9_00_HD_CTRY_OF_DOMICILE  \
0                     0                        0                      None   

   F9_00_HD_EXEMPT_STATUS_4847A1  F9_00_HD_EXEMPT_STATUS_501C  \
0                              0                         <NA>   

   F9_00_HD_EXEMPT_STATUS_501C3  F9_00_HD_FINAL_RETURN  \
0                             1                      0   

  F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  F9_00_HD_GROUP_RETURN  \
0                      None               376800                      0   

   F9_00_HD_INCLUDES_SUBORD_ORGS  F9_00_HD_INITIAL_RETURN  \
0                              0                        0   

  F9_00_HD_PRIN_OFF_NAME F9_00_HD_SIGNING_OFFICER_SIGNTR  \
0            DANIEL FORD                      2023-11-13   

  F9_00_HD_SPECIAL_CONDITION_DESC F9_00_HD_STATE_OF_DOMICILE  \
0                            None                         ME   

  F9_00_HD_TAX_PER_BEGIN F9_00_HD_TAX_PER_END  F9_00_HD_TAX_YEAR  \
0             2022-01-01           2022-12-31               2022   

        F9_00_HD_TIME_STAMP  F9_00_HD_TYPE_ORG_ASSOCIATION  \
0 2023-11-14 16:30:26+00:00                              0   

   F9_00_HD_TYPE_ORG_CORP  F9_00_HD_TYPE_ORG_OTHER  \
0                       1                        0   

  F9_00_HD_TYPE_ORG_OTHER_DESC  F9_00_HD_TYPE_ORG_TRUST F9_00_HD_WEBSITE  \
0                         None                        0  WWW.AYCSAIL.ORG   

   F9_00_HD_YEAR_FORMED  F9_01_PC_BEN_PAID_MEMB_PRIOR  \
0                  1937                             0   

   F9_01_PC_CONTR_GRANTS_CURR  F9_01_PC_CONTR_GRANTS_PRIOR  \
0                      279970                            0   

   F9_01_PC_GRANTS_PRIOR  F9_01_PC_INDEP_VOTING_MEMB  \
0                      0                          13   

   F9_01_PC_INVEST_INCOME_PRIOR  F9_01_PC_NET_ASSETS_BOY  \
0                             0                   273331   

   F9_01_PC_OTHER_EXPENSE_PRIOR  F9_01_PC_OTHER_REV_PRIOR  \
0                             0                         0   

   F9_01_PC_PROF_FUNDRISING_EXP_CURR  F9_01_PC_PROF_FUNDRISING_EXP_PRIOR  \
0                                  0                                   0   

   F9_01_PC_PROG_SERVICE_REV_PRIOR  F9_01_PC_REV_LESS_EXP_CURR  \
0                                0                      184620   

   F9_01_PC_REV_LESS_EXP_PRIOR  F9_01_PC_TERMINATION_CONTRACTION  \
0                            0                                 0   

   F9_01_PC_TOT_ASSETS_EOY  F9_01_PC_TOT_EXP_PRIOR  \
0                   413907                       0   

   F9_01_PC_TOT_FNDR_EXP_CURR  F9_01_PC_TOT_INDIV_EMPLOYED  \
0                        2744                           16   

   F9_01_PC_TOT_INDIV_VOLUNTE

### Collapse data for the duplicates, taking the last entry for each org/year row
This works because we have (as also done earlier) sorted the dataset by tax year and signing officer date

In [26]:
%%time
df = df.sort_values(['EIN', 'F9_00_HD_TAX_YEAR', 'F9_00_HD_SIGNING_OFFICER_SIGNTR'])
df[:1]

CPU times: total: 6.58 s
Wall time: 6.75 s


_id OrganizationName  \
0  65c1a1d52a9ba8ce45342904             None   

                                                                   URL   DLN  \
0  https://s3.amazonaws.com/irs-form-990/202323189349305317_public.xml  None   

  TaxPeriod  F9_09_PC_FEES_FOR_SVCE_FR_TOT F9_00_HD_BUILD_TIME_STAMP  \
0      None                              0 2023-04-26 12:10:37+00:00   

  fiscal_year        EIN  Name NameControl Phone  \
0        2022  010017496  None        None  None   

                                                                                                                                                                                                                      USAddress  \
0  {'AddressLine1': None, 'AddressLine1Txt': 'PO BOX 534', 'AddressLine2': None, 'AddressLine2Txt': None, 'City': None, 'CityNm': 'YORK HARBOR', 'State': None, 'StateAbbreviationCd': 'ME', 'ZIPCd': '03911', 'ZIPCode': None}   

  ForeignAddress InCareOfName  \
0           None         None   

                                             BusinessName  \
0  {'BusinessNameLine1Txt': 'AGAMENTICUS YACHT CLUB INC'}   

  BusinessNameControlTxt    PhoneNum InCareOfNm ForeignPhoneNum  \
0                   AGAM  2073638510       None            None   

   F9_00_HD_ADDR_CHANGE  F9_00_HD_AMENDED_RETURN F9_00_HD_CTRY_OF_DOMICILE  \
0                     0                        0                      None   

   F9_00_HD_EXEMPT_STATUS_4847A1  F9_00_HD_EXEMPT_STATUS_501C  \
0                              0                         <NA>   

   F9_00_HD_EXEMPT_STATUS_501C3  F9_00_HD_FINAL_RETURN  \
0                             1                      0   

  F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  F9_00_HD_GROUP_RETURN  \
0                      None               376800                      0   

   F9_00_HD_INCLUDES_SUBORD_ORGS  F9_00_HD_INITIAL_RETURN  \
0                              0                        0   

  F9_00_HD_PRIN_OFF_NAME F9_00_HD_SIGNING_OFFICER_SIGNTR  \
0            DANIEL FORD                      2023-11-13   

  F9_00_HD_SPECIAL_CONDITION_DESC F9_00_HD_STATE_OF_DOMICILE  \
0                            None                         ME   

  F9_00_HD_TAX_PER_BEGIN F9_00_HD_TAX_PER_END  F9_00_HD_TAX_YEAR  \
0             2022-01-01           2022-12-31               2022   

        F9_00_HD_TIME_STAMP  F9_00_HD_TYPE_ORG_ASSOCIATION  \
0 2023-11-14 16:30:26+00:00                              0   

   F9_00_HD_TYPE_ORG_CORP  F9_00_HD_TYPE_ORG_OTHER  \
0                       1                        0   

  F9_00_HD_TYPE_ORG_OTHER_DESC  F9_00_HD_TYPE_ORG_TRUST F9_00_HD_WEBSITE  \
0                         None                        0  WWW.AYCSAIL.ORG   

   F9_00_HD_YEAR_FORMED  F9_01_PC_BEN_PAID_MEMB_PRIOR  \
0                  1937                             0   

   F9_01_PC_CONTR_GRANTS_CURR  F9_01_PC_CONTR_GRANTS_PRIOR  \
0                      279970                            0   

   F9_01_PC_GRANTS_PRIOR  F9_01_PC_INDEP_VOTING_MEMB  \
0                      0                          13   

   F9_01_PC_INVEST_INCOME_PRIOR  F9_01_PC_NET_ASSETS_BOY  \
0                             0                   273331   

   F9_01_PC_OTHER_EXPENSE_PRIOR  F9_01_PC_OTHER_REV_PRIOR  \
0                             0                         0   

   F9_01_PC_PROF_FUNDRISING_EXP_CURR  F9_01_PC_PROF_FUNDRISING_EXP_PRIOR  \
0                                  0                                   0   

   F9_01_PC_PROG_SERVICE_REV_PRIOR  F9_01_PC_REV_LESS_EXP_CURR  \
0                                0                      184620   

   F9_01_PC_REV_LESS_EXP_PRIOR  F9_01_PC_TERMINATION_CONTRACTION  \
0                            0                                 0   

   F9_01_PC_TOT_ASSETS_EOY  F9_01_PC_TOT_EXP_PRIOR  \
0                   413907                       0   

   F9_01_PC_TOT_FNDR_EXP_CURR  F9_01_PC_TOT_INDIV_EMPLOYED  \
0                        2744                           16   

   F9_01_PC_TOT_INDIV_VOLUNTE

In [27]:
df[(df['EIN'].isin(dups))&(df['number of org-year duplicates'].notnull())][['URL', 'BMF_ORG_NAME_CURRENT', 'F9_00_HD_TAX_YEAR', 
                          'F9_00_HD_BUILD_TIME_STAMP', 'F9_00_HD_TIME_STAMP', 'F9_00_HD_SIGNING_OFFICER_SIGNTR', 'F9_00_HD_AMENDED_RETURN', 'number of org-year duplicates']][:10]

URL  \
243  https://s3.amazonaws.com/irs-form-990/201910169349300336_public.xml   
242  https://s3.amazonaws.com/irs-form-990/201900449349300005_public.xml   
290  https://s3.amazonaws.com/irs-form-990/201612149349301391_public.xml   
289  https://s3.amazonaws.com/irs-form-990/201612289349306266_public.xml   
292  https://s3.amazonaws.com/irs-form-990/202032129349300713_public.xml   
293  https://s3.amazonaws.com/irs-form-990/202043149349303484_public.xml   
295  https://s3.amazonaws.com/irs-form-990/202132259349300308_public.xml   
294  https://s3.amazonaws.com/irs-form-990/202231649349301203_public.xml   
359  https://s3.amazonaws.com/irs-form-990/201731359349301753_public.xml   
360  https://s3.amazonaws.com/irs-form-990/201820089349300342_public.xml   

                      BMF_ORG_NAME_CURRENT  F9_00_HD_TAX_YEAR  \
243                  BANGOR CHILDRENS HOME               2017   
242                  BANGOR CHILDRENS HOME               2017   
290                  MT DESERT ISLAND YMCA               2015   
289                  MT DESERT ISLAND YMCA               2015   
292                  MT DESERT ISLAND YMCA               2018   
293                  MT DESERT ISLAND YMCA               2018   
295                  MT DESERT ISLAND YMCA               2019   
294                  MT DESERT ISLAND YMCA               2019   
359  PRESIDENT & TRUSTEES OF COLBY COLLEGE               2015   
360  PRESIDENT & TRUSTEES OF COLBY COLLEGE               2015   

    F9_00_HD_BUILD_TIME_STAMP       F9_00_HD_TIME_STAMP  \
243 2019-02-21 02:37:17+00:00 2019-01-16 21:02:25+00:00   
242 2019-02-21 02:37:17+00:00 2019-02-13 13:49:42+00:00   
290 2016-12-15 16:53:06+00:00 2016-08-01 20:26:21+00:00   
289 2016-09-27 15:27:22+00:00 2016-08-15 21:36:56+00:00   
292 2020-09-22 18:59:49+00:00 2020-07-30 18:49:28+00:00   
293 2021-01-29 14:40:06+00:00 2020-11-09 21:14:55+00:00   
295 2022-09-23 18:48:47+00:00 2021-08-13 12:25:24+00:00   
294 2022-09-23 18:48:47+00:00 2022-06-13 17:50:45+00:00   
359 2017-02-10 21:41:12+00:00 2017-05-10 23:52:12+00:00   
360 2018-03-14 21:41:22+00:00 2017-12-18 16:05:20+00:00   

    F9_00_HD_SIGNING_OFFICER_SIGNTR  F9_00_HD_AMENDED_RETURN  \
243                      2019-01-16                        0   
242                      2019-02-06                        1   
290                      2016-08-01                        0   
289                      2016-08-15                        1   
292                      2020-07-30                        0   
293                      2020-11-03                        1   
295                      2021-08-11                        0   
294                      2022-06-09                        1   
359                      2017-03-18                        0   
360                      2017-12-18                        1   

     number of org-year duplicates  
243                           2.00  
242                           2.00  
290                           2.00  
289                           2.00  
292                           2.00  
293                           2.00  
295                           2.00  
294                           2.00  
359                           2.00  
360                           2.00

In [41]:
len(grouped)

33668

In [34]:
df['number of org-year duplicates'].value_counts()

number of org-year duplicates
2.00    65328
3.00     2697
4.00      360
5.00       70
6.00        6
Name: count, dtype: int64

In [35]:
df['number of org-year duplicates'].value_counts().sum()

68461

In [36]:
df['number of org-year duplicates'].describe()

count   68461.00
mean        2.05
std         0.26
min         2.00
25%         2.00
50%         2.00
75%         2.00
max         6.00
Name: number of org-year duplicates, dtype: float64

In [37]:
print(len(df[df['number of org-year duplicates'].notnull()]))
print(len(set(df[df['number of org-year duplicates'].notnull()]['EIN'].tolist())))

68461
26466


In [38]:
len(df[(df['EIN'].isin(dups))&(df['number of org-year duplicates'].notnull())])

68461

In [42]:
dups = df[df['number of org-year duplicates'].notnull()]['EIN'].tolist()
print(len(dups), len(set(dups)))

68461 26466


<br>Of the 68,461 filings with duplicates (for 26,466 unique EINs), there are 33,668 affected *EIN filing/years*.

Note that the code block below takes the `last( )` duplicate. This is why the descriptives for `number of org-year duplicates` are different than they are in `df`

Given that there are 68,461 filings in `df` with a duplicate value, and that of these there are 33,668 'valid' filings, we should expect the final dataset to have 2,598,477 observations.

In [43]:
#11/14/2021 -- NOT RUN
print(len(df))
print(len(df) - 68461)
print(len(df) - 68461 + 33668)

2633270
2564809
2598477


In [44]:
gc.collect()

0

In [45]:
%%time
df = df.groupby(['EIN', 'F9_00_HD_TAX_YEAR']).last()
print(len(df))
df = df.reset_index()
df[:2]

2598477
CPU times: total: 36.4 s
Wall time: 37.3 s


EIN  F9_00_HD_TAX_YEAR                       _id  \
0  010017496               2022  65c1a1d52a9ba8ce45342904   
1  010024645               2009  5d01baf378ffca27b4286d51   

            OrganizationName  \
0                       None   
1  BANGOR SYMPHONY ORCHESTRA   

                                                                   URL  \
0  https://s3.amazonaws.com/irs-form-990/202323189349305317_public.xml   
1  https://s3.amazonaws.com/irs-form-990/201121379349300422_public.xml   

              DLN TaxPeriod  F9_09_PC_FEES_FOR_SVCE_FR_TOT  \
0            None      None                              0   
1  93493137004221    201006                              0   

  F9_00_HD_BUILD_TIME_STAMP fiscal_year  \
0 2023-04-26 12:10:37+00:00        2022   
1 2016-02-24 21:20:13+00:00        None   

                                                 Name NameControl       Phone  \
0                                                None        None        None   
1  {'BusinessNameLine1': 'Bangor Symphony Orchestra'}        BANG  2079425555   

                                                                                                                                                                                                                      USAddress  \
0  {'AddressLine1': None, 'AddressLine1Txt': 'PO BOX 534', 'AddressLine2': None, 'AddressLine2Txt': None, 'City': None, 'CityNm': 'YORK HARBOR', 'State': None, 'StateAbbreviationCd': 'ME', 'ZIPCd': '03911', 'ZIPCode': None}   
1  {'AddressLine1': 'PO Box 1441', 'AddressLine1Txt': None, 'AddressLine2': None, 'AddressLine2Txt': None, 'City': 'Bangor', 'CityNm': None, 'State': 'ME', 'StateAbbreviationCd': None, 'ZIPCd': None, 'ZIPCode': '044021441'}   

  ForeignAddress InCareOfName  \
0           None         None   
1           None         None   

                                             BusinessName  \
0  {'BusinessNameLine1Txt': 'AGAMENTICUS YACHT CLUB INC'}   
1                                                    None   

  BusinessNameControlTxt    PhoneNum InCareOfNm ForeignPhoneNum  \
0                   AGAM  2073638510       None            None   
1                   None        None       None            None   

   F9_00_HD_ADDR_CHANGE  F9_00_HD_AMENDED_RETURN F9_00_HD_CTRY_OF_DOMICILE  \
0                     0                        0                      None   
1                     0                        0                      None   

   F9_00_HD_EXEMPT_STATUS_4847A1  F9_00_HD_EXEMPT_STATUS_501C  \
0                              0                         <NA>   
1                              0                            3   

   F9_00_HD_EXEMPT_STATUS_501C3  F9_00_HD_FINAL_RETURN  \
0                             1                      0   
1                             0                      0   

  F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  F9_00_HD_GROUP_RETURN  \
0                      None               376800                      0   
1                      None              1080566                      0   

   F9_00_HD_INCLUDES_SUBORD_ORGS  F9_00_HD_INITIAL_RETURN  \
0                              0                        0   
1                              0                        0   

  F9_00_HD_PRIN_OFF_NAME F9_00_HD_SIGNING_OFFICER_SIGNTR  \
0            DANIEL FORD                      2023-11-13   
1              Lee Chick                      2011-05-17   

  F9_00_HD_SPECIAL_CONDITION_DESC F9_00_HD_STATE_OF_DOMICILE  \
0                            None                         ME   
1                            None                         ME   

  F9_00_HD_TAX_PER_BEGIN F9_00_HD_TAX_PER_END       F9_00_HD_TIME_STAMP  \
0             2022-01-01           2022-12-31 2023-11-14 16:30:26+00:00   
1             2009-07-01           2010-06-30 2011-05-16 15:26:38+00:00   

   F9_00_HD_TYPE_ORG_ASSOCIATION  F9_00_HD_TYPE_ORG_CORP  \
0                              0                       1   
1                              0        

#### Now inspect dataset

In [46]:
%%time
df[(df['EIN'].isin(dups))&(df['number of org-year duplicates']>1)][['URL', 'BMF_ORG_NAME_CURRENT', 'F9_00_HD_TAX_YEAR', 
                          'F9_00_HD_BUILD_TIME_STAMP', 'F9_00_HD_TIME_STAMP', 'F9_00_HD_SIGNING_OFFICER_SIGNTR', 'F9_00_HD_AMENDED_RETURN', 'number of org-year duplicates']][:10]

CPU times: total: 516 ms
Wall time: 556 ms


URL  \
243   https://s3.amazonaws.com/irs-form-990/201900449349300005_public.xml   
289   https://s3.amazonaws.com/irs-form-990/201612289349306266_public.xml   
291   https://s3.amazonaws.com/irs-form-990/202043149349303484_public.xml   
292   https://s3.amazonaws.com/irs-form-990/202231649349301203_public.xml   
359   https://s3.amazonaws.com/irs-form-990/201820089349300342_public.xml   
546   https://s3.amazonaws.com/irs-form-990/201922599349301132_public.xml   
617   https://s3.amazonaws.com/irs-form-990/202332149349300838_public.xml   
666   https://s3.amazonaws.com/irs-form-990/202140199349300314_public.xml   
1078  https://s3.amazonaws.com/irs-form-990/201412129349300711_public.xml   
1178  https://s3.amazonaws.com/irs-form-990/201400439349301280_public.xml   

                       BMF_ORG_NAME_CURRENT  F9_00_HD_TAX_YEAR  \
243                   BANGOR CHILDRENS HOME               2017   
289                   MT DESERT ISLAND YMCA               2015   
291                   MT DESERT ISLAND YMCA               2018   
292                   MT DESERT ISLAND YMCA               2019   
359   PRESIDENT & TRUSTEES OF COLBY COLLEGE               2015   
546                          MERCY HOSPITAL               2017   
617     COMMUNITY HEALTH & NURSING SERVICES               2021   
666                        WAYNFLETE SCHOOL               2018   
1078                  BANGOR HUMANE SOCIETY               2013   
1178                      WAYFINDER SCHOOLS               2012   

     F9_00_HD_BUILD_TIME_STAMP       F9_00_HD_TIME_STAMP  \
243  2019-02-21 02:37:17+00:00 2019-02-13 13:49:42+00:00   
289  2016-09-27 15:27:22+00:00 2016-08-15 21:36:56+00:00   
291  2021-01-29 14:40:06+00:00 2020-11-09 21:14:55+00:00   
292  2022-09-23 18:48:47+00:00 2022-06-13 17:50:45+00:00   
359  2018-03-14 21:41:22+00:00 2017-12-18 16:05:20+00:00   
546  2019-02-21 02:37:17+00:00 2019-09-16 16:33:06+00:00   
617  2023-04-26 12:10:37+00:00 2023-08-02 16:44:43+00:00   
666  2021-01-29 14:40:06+00:00 2021-01-19 18:44:54+00:00   
1078 2015-11-30 17:44:51+00:00 2014-07-31 14:32:48+00:00   
1178 2016-03-07 17:11:31+00:00 2014-02-12 19:23:29+00:00   

     F9_00_HD_SIGNING_OFFICER_SIGNTR  F9_00_HD_AMENDED_RETURN  \
243                       2019-02-06                        1   
289                       2016-08-15                        1   
291                       2020-11-03                        1   
292                       2022-06-09                        1   
359                       2017-12-18                        1   
546                       2019-09-16                        1   
617                       2023-07-05                        0   
666                       2021-01-19                        1   
1078                      2014-07-31                        1   
1178                      2014-02-12                        1   

      number of org-year duplicates  
243                            2.00  
289                            2.00  
291                            2.00  
292                            2.00  
359                            2.00  
546                            2.00  
617                            2.00  
666                            2.00  
1078                           2.00  
1178                           2.00

#### Look for duplicates again
There are no more duplicates

In [49]:
%%time
grouped = df.groupby(['EIN', 'F9_00_HD_TAX_YEAR']).count()
print(len(grouped[grouped['URL']>1]))
print(grouped[grouped['URL']>1]['URL'].value_counts())
grouped = grouped[grouped['URL']>1]
grouped = grouped.reset_index()
grouped = grouped[['EIN', 'F9_00_HD_TAX_YEAR', 'URL']]
grouped.columns = ['EIN', 'F9_00_HD_TAX_YEAR', 'number of org-year duplicates']
print(len(grouped))
grouped[:2]

0
Series([], Name: count, dtype: int64)
0
CPU times: total: 10.5 s
Wall time: 10.9 s


Empty DataFrame
Columns: [EIN, F9_00_HD_TAX_YEAR, number of org-year duplicates]
Index: []

#### Create variable that indicates whether there was a duplicate

In [50]:
df['filing_year_had_duplicate'] = np.where(df['number of org-year duplicates']>0, 1,0)
df['filing_year_had_duplicate'].value_counts()

filing_year_had_duplicate
0    2564809
1      33668
Name: count, dtype: int64

In [51]:
df['number of org-year duplicates'].value_counts()

number of org-year duplicates
2.00    32664
3.00      899
4.00       90
5.00       14
6.00        1
Name: count, dtype: int64

In [52]:
pd.crosstab(df['filing_year_had_duplicate'], df['number of org-year duplicates'])

number of org-year duplicates   2.00  3.00  4.00  5.00  6.00
filing_year_had_duplicate                                   
1                              32664   899    90    14     1

In [53]:
df = df.drop('number of org-year duplicates', axis=1)

#### Inspect Data

In [54]:
print('# of columns:', len(df.columns))
print('# of observations:', len(df))

# of columns: 357
# of observations: 2598477


In [55]:
df.sample(1)

EIN  F9_00_HD_TAX_YEAR                       _id  \
1558567  520610545               2019  617c5978b1ca7b56cbd8f4e9   

              OrganizationName  \
1558567  SUBURBAN HOSPITAL INC   

                                                                         URL  \
1558567  https://s3.amazonaws.com/irs-form-990/202111339349303186_public.xml   

                    DLN TaxPeriod  F9_09_PC_FEES_FOR_SVCE_FR_TOT  \
1558567  93493133031861    202006                              0   

        F9_00_HD_BUILD_TIME_STAMP fiscal_year  Name NameControl Phone  \
1558567 2021-01-29 14:40:06+00:00        2020  None        None  None   

                                                                                                                                                                                                                                           USAddress  \
1558567  {'AddressLine1': None, 'AddressLine1Txt': '8600 OLD GEORGETOWN ROAD', 'AddressLine2': None, 'AddressLine2Txt': None, 'City': None, 'CityNm': 'BETHESDA', 'State': None, 'StateAbbreviationCd': 'MD', 'ZIPCd': '208141497', 'ZIPCode': None}   

        ForeignAddress InCareOfName  \
1558567           None         None   

                                              BusinessName  \
1558567  {'BusinessNameLine1Txt': 'SUBURBAN HOSPITAL INC'}   

        BusinessNameControlTxt    PhoneNum InCareOfNm ForeignPhoneNum  \
1558567                   SUBU  4439975771       None            None   

         F9_00_HD_ADDR_CHANGE  F9_00_HD_AMENDED_RETURN  \
1558567                     0                        0   

        F9_00_HD_CTRY_OF_DOMICILE  F9_00_HD_EXEMPT_STATUS_4847A1  \
1558567                      None                              0   

         F9_00_HD_EXEMPT_STATUS_501C  F9_00_HD_EXEMPT_STATUS_501C3  \
1558567                         <NA>                             1   

         F9_00_HD_FINAL_RETURN F9_00_HD_GROSS_EXEMPT_NUM  F9_00_HD_GROSS_RCPT  \
1558567                      0                      None            354492109   

         F9_00_HD_GROUP_RETURN  F9_00_HD_INCLUDES_SUBORD_ORGS  \
1558567                      0                              0   

         F9_00_HD_INITIAL_RETURN F9_00_HD_PRIN_OFF_NAME  \
1558567                        0           MARTIN BASSO   

        F9_00_HD_SIGNING_OFFICER_SIGNTR F9_00_HD_SPECIAL_CONDITION_DESC  \
1558567                      2021-05-13                            None   

        F9_00_HD_STATE_OF_DOMICILE F9_00_HD_TAX_PER_BEGIN  \
1558567                         MD             2019-07-01   

        F9_00_HD_TAX_PER_END       F9_00_HD_TIME_STAMP  \
1558567           2020-06-30 2021-05-13 18:40:00+00:00   

         F9_00_HD_TYPE_ORG_ASSOCIATION  F9_00_HD_TYPE_ORG_CORP  \
1558567                              0                       1   

         F9_00_HD_TYPE_ORG_OTHER F9_00_HD_TYPE_ORG_OTHER_DESC  \
1558567                        0                         None   

         F9_00_HD_TYPE_ORG_TRUST          F9_00_HD_WEBSITE  \
1558567                        0  WWW.SUBURBANHOSPITAL.ORG   

         F9_00_HD_YEAR_FORMED  F9_01_PC_BEN_PAID_MEMB_PRIOR  \
1558567                  1942                             0   

         F9_01_PC_CONTR_GRANTS_CURR  F9_01_PC_CONTR_GRANTS_PRIOR  \
1558567                    23143939                      3514459   

         F9_01_PC_GRANTS_PRIOR  F9_01_PC_INDEP_VOTING_MEMB  \
1558567                  92511                          17   

         F9_01_PC_INVEST_INCOME_PRIOR  F9_01_PC_NET_ASSETS_BOY  \
1558567                      22707990                401174828   

         F9_01_PC_OTHER_EXPENSE_PRIOR  F9_01_PC_OTHER_REV_PRIOR  \
1558567                     163343243                  14987211   

         F9_01_PC_PROF_FUNDRISING_EXP_CURR  \
1558567                                  0   

         F9_01_PC_PROF_FUNDRISING_EXP_PRIOR  F9_01_PC_PROG_SERVICE_REV_PRIOR  \
1558567                                   0                        289980427   

         F9_01_PC_R

In [56]:
print(df.columns.tolist())

['EIN', 'F9_00_HD_TAX_YEAR', '_id', 'OrganizationName', 'URL', 'DLN', 'TaxPeriod', 'F9_09_PC_FEES_FOR_SVCE_FR_TOT', 'F9_00_HD_BUILD_TIME_STAMP', 'fiscal_year', 'Name', 'NameControl', 'Phone', 'USAddress', 'ForeignAddress', 'InCareOfName', 'BusinessName', 'BusinessNameControlTxt', 'PhoneNum', 'InCareOfNm', 'ForeignPhoneNum', 'F9_00_HD_ADDR_CHANGE', 'F9_00_HD_AMENDED_RETURN', 'F9_00_HD_CTRY_OF_DOMICILE', 'F9_00_HD_EXEMPT_STATUS_4847A1', 'F9_00_HD_EXEMPT_STATUS_501C', 'F9_00_HD_EXEMPT_STATUS_501C3', 'F9_00_HD_FINAL_RETURN', 'F9_00_HD_GROSS_EXEMPT_NUM', 'F9_00_HD_GROSS_RCPT', 'F9_00_HD_GROUP_RETURN', 'F9_00_HD_INCLUDES_SUBORD_ORGS', 'F9_00_HD_INITIAL_RETURN', 'F9_00_HD_PRIN_OFF_NAME', 'F9_00_HD_SIGNING_OFFICER_SIGNTR', 'F9_00_HD_SPECIAL_CONDITION_DESC', 'F9_00_HD_STATE_OF_DOMICILE', 'F9_00_HD_TAX_PER_BEGIN', 'F9_00_HD_TAX_PER_END', 'F9_00_HD_TIME_STAMP', 'F9_00_HD_TYPE_ORG_ASSOCIATION', 'F9_00_HD_TYPE_ORG_CORP', 'F9_00_HD_TYPE_ORG_OTHER', 'F9_00_HD_TYPE_ORG_OTHER_DESC', 'F9_00_HD_TYPE_ORG_

In [57]:
df.describe().T

count    mean       std       min     25%  \
F9_00_HD_TAX_YEAR             2598477.00 2017.37      3.89   2009.00 2014.00   
F9_09_PC_FEES_FOR_SVCE_FR_TOT 2598477.00 4948.57 158234.29 -35000.00    0.00   
F9_00_HD_ADDR_CHANGE          2598477.00    0.04      0.20      0.00    0.00   
F9_00_HD_AMENDED_RETURN       2598477.00    0.01      0.11      0.00    0.00   
F9_00_HD_EXEMPT_STATUS_4847A1 2598477.00    0.00      0.00      0.00    0.00   
...                                  ...     ...       ...       ...     ...   
BMF_ORG_YEAR_FIRST            2594238.00 1999.07     10.20   1989.00 1989.00   
BMF_ORG_YEAR_LAST             2594238.00 2023.76      1.21   1989.00 2024.00   
BMF_ORG_YEAR_COUNT            2594238.00   25.68     10.28      1.00   17.00   
BMF_ORG_FISCAL_PERIOD         2594238.00    6.38      3.38      1.00    3.00   
filing_year_had_duplicate     2598477.00    0.01      0.11      0.00    0.00   

                                  50%     75%         max  
F9_00_HD_TAX_YEAR             2018.00 2021.00     2024.00  
F9_09_PC_FEES_FOR_SVCE_FR_TOT    0.00    0.00 97665783.00  
F9_00_HD_ADDR_CHANGE             0.00    0.00        1.00  
F9_00_HD_AMENDED_RETURN          0.00    0.00        1.00  
F9_00_HD_EXEMPT_STATUS_4847A1    0.00    0.00        0.00  
...                               ...     ...         ...  
BMF_ORG_YEAR_FIRST            1997.00 2007.00     2024.00  
BMF_ORG_YEAR_LAST             2024.00 2024.00     2024.00  
BMF_ORG_YEAR_COUNT              28.00   36.00       36.00  
BMF_ORG_FISCAL_PERIOD            6.00    9.00       12.00  
filing_year_had_duplicate        0.00    0.00        1.00  

[288 rows x 8 columns]

In [58]:
df['F9_00_HD_TAX_YEAR'].value_counts().sort_index()

F9_00_HD_TAX_YEAR
2009     28060
2010     91891
2011    119309
2012    134445
2013    149089
2014    163652
2015    175038
2016    182842
2017    191289
2018    200139
2019    213238
2020    237976
2021    251105
2022    259785
2023    198193
2024      2426
Name: count, dtype: Int64

In [59]:
len(set(df['EIN'].tolist()))

351875

In [60]:
print('# of columns:', len(df.columns))
print('# of observations:', len(df))

# of columns: 357
# of observations: 2598477


In [61]:
df['501c3'].value_counts()

501c3
1    2598477
Name: count, dtype: int64

In [62]:
len(df)

2598477

In [63]:
len(set(df['EIN']))

351875

#### Save DF

In [64]:
%%time
print ("Current date and time : ", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"), '\n')
df.to_feather('D:/990_and_bmf_april_2025_all_controls_351875_orgs_2598477_filings_no_duplicates.feather')

Current date and time :  2025-04-20 19:59:36 

CPU times: total: 30.3 s
Wall time: 25.4 s


In [65]:
%%time
print ("Current date and time : ", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"), '\n')
df.to_parquet("D:/990_and_bmf_april_2025_all_controls_351875_orgs_2598477_filings_no_duplicates.parquet", engine="pyarrow", compression="snappy", index=False)

Current date and time :  2025-04-20 20:00:01 

CPU times: total: 1min 8s
Wall time: 1min 7s


In [66]:
%%time
print ("Current date and time : ", datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"), '\n')
df.to_pickle('990_and_bmf_april_2025_all_controls_351875_orgs_2598477_filings_no_duplicates.pkl.gz', compression='gzip')

Current date and time :  2025-04-20 20:01:09 

CPU times: total: 34min 34s
Wall time: 35min 14s
